elements/drag-drop-elements/pl-drag-drop-element.py

```diff
@@ -0,0 +1,367 @@
+
+    student_answer = data['submitted_answers'][answerName]['student_raw_submission']
+    student_answer_indent = data['submitted_answers'][answerName]['student_answer_indent']
+    permutationMode = pl.get_string_attrib(element, 'permutation-mode', 'html-order')
+    true_answer = data['correct_answers'][answerName]['correct_answers']
+    true_answer_indent = data['correct_answers'][answerName]['correct_answers_indent']
+
+    indent_score = 0
+    final_score = 0
+    feedback = ''
+
+    if permutationMode == 'any':
+        intersection = list(set(student_answer) & set(true_answer))
+        final_score = float(len(intersection) / len(true_answer))
+    elif permutationMode == 'html-order':
+        final_score = 1.0 if student_answer == true_answer else 0.0
+    elif permutationMode == 'ranking':
+        ranking = data['submitted_answers'][answerName]['student_submission_ordering']
+        correctness = 1
+        partial_credit = 0
+        if len(ranking) != 0 and len(ranking) == len(true_answer):
+            if ranking[0] == 1:
+                partial_credit = 1  # student will at least get 1 point for getting first element correct
+            for x in range(0, len(ranking) - 1):
+                if int(ranking[x]) == -1:
+                    correctness = 0
+                    break
+                if int(ranking[x]) <= int(ranking[x + 1]):
+                    correctness += 1
+                else:
+                    correctness = 0
+                    break
+        else:
+            correctness = 0
+        correctness = max(correctness, partial_credit)
+        final_score = float(correctness / len(true_answer))
+
+    check_indentation = pl.get_string_attrib(element, 'check-indentation', 'false')
+    # check indents, and apply penalty if applicable
+    if true_answer_indent.count('-1') != len(true_answer_indent) or check_indentation == 'true':
+        for i, indent in enumerate(student_answer_indent):
+            if indent == true_answer_indent[i] or true_answer_indent[i] == '-1':
+                indent_score += 1
+        final_score = final_score * (indent_score / len(true_answer_indent))
+
+    data['partial_scores'][answerName] = {'score': final_score, 'feedback': feedback}
+
+
+def test(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    answerName = pl.get_string_attrib(element, 'answers-name')
+    answerNameField = answerName + '-input'
+
+    # incorrect and correct answer test cases
+    # this creates the EXPECTED SUBMISSION field for test cases
+    if data['test_type'] == 'correct':
+        temp = data['correct_answers'][answerName]['correct_answers'].copy()  # temp array to hold the correct answers
+        true_answer_indent = data['correct_answers'][answerName]['correct_answers_indent']
+        for index, answer in enumerate(temp):
+            temp[index] = answer + ':::' + true_answer_indent[index]
```

Comment:
It would be nice to get rid of the strange delimiters in this function too.

Reply:
oh this is the testing function that I forgot to update now that we're using JSON

Reply:
updated in the latest commit